sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderUtil.java

```diff
@@ -67,18 +68,28 @@ private static void logFilteredOutPath(ThrowStatementVertex throwStatementVertex
         }
     }
 
-    /**
-     * The {@code ApexPathCollapser} that keeps track of all paths and attempts to collapse them
-     * whenever a forked method has returned a result.
-     */
-    private final ApexPathCollapser apexPathCollapser;
+    /** Used to give each object a unique id */
+    private static final AtomicLong ID_GENERATOR = new AtomicLong();
+
+    private final Long pathExpansionId;
+
+    private final PathExpansionRegistry registry;
 
-    private ApexPathExpanderUtil(ApexPathExpanderConfig config) {
+    private ApexPathExpanderUtil(ApexPathExpanderConfig config, PathExpansionRegistry registry) {
+        this.pathExpansionId = ID_GENERATOR.incrementAndGet();
+        this.registry = registry;
+
+        // Create and register ApexPathCollapser
+        final ApexPathCollapser apexPathCollapser;
         if (config.getDynamicCollapsers().isEmpty()) {
-            this.apexPathCollapser = NoOpApexPathCollapser.getInstance();
+            apexPathCollapser = new NoOpApexPathCollapser(pathExpansionId);
```

Comment:
This is slightly controversial. I've converted a singleton `NoOpApexPathCollapser` into a normal class. Had to make this to allow registry lookup of any type of `ApexPathCollapser`. This implementation is pretty light as well. However, I'm now reconsidering this along with my other comment on making `ApexPathCollapserImpl` indexable.

Reply:
The way you've done it looks sound, but I agree that we should definitely determine whether it actually has a beneficial impact before we keep the change.

Reply:
This appeared to have some beneficial effect when I started, but now that I have more information, I might've been incorrect initially.